Sources/NIOCore/ByteBuffer-lengthPrefix.swift

```diff
@@ -0,0 +1,105 @@
+        var totalBytesWritten = 0
+        
+        let lengthPrefixIndex = self.writerIndex
+        // Write a zero as a placeholder which will later be overwritten by the actual number of bytes written
+        totalBytesWritten += self.writeInteger(.zero, endianness: endianness, as: Integer.self)
+        
+        let startWriterIndex = writerIndex
+        let messageLength = try writeMessage(&self)
+        let endWriterIndex = writerIndex
+        
+        totalBytesWritten += messageLength
+        
+        let actualBytesWritten = endWriterIndex - startWriterIndex
+        assert(
+            actualBytesWritten == messageLength, 
+            "writeMessage returned \(messageLength) bytes, but actually \(actualBytesWritten) bytes were written, but they should be the same"
+        )
+        
+        guard let lengthPrefix = Integer(exactly: messageLength) else {
+            throw LengthPrefixError.messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat
+        }
+        
+        self.setInteger(lengthPrefix, at: lengthPrefixIndex, endianness: endianness, as: Integer.self)
+        
+        return totalBytesWritten
+    }
+}
+
+extension ByteBuffer {
+    /// Reads an `integer` from `self` and passes it to `readMessage`. 
+    /// It is checked that exactly the number of bytes specified in the length prefix is read during the call to `readMessage`. 
+    /// 
+    /// If nil is returned, `readerIndex` is **not** moved forward.
+    /// - Parameters:
+    ///     - endianness: The endianness of the length prefix `Integer` in this `ByteBuffer` (defaults to big endian).
+    ///     - integer: the desired `Integer` type used to read the length prefix
+    ///     - readMessage: A closure that takes the message length, reads exactly that number of bytes from the given `ByteBuffer` and returns the result.
+    /// - Returns: `nil` if the length prefix could not be read, 
+    ///            the length prefix is negative or
+    ///            the buffer does not contain enough bytes to read a message of this length.
+    ///            Otherwise the result of `readMessage`.
+    @inlinable
+    public mutating func readLengthPrefixed<Integer, Result>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        readMessage: (Int, inout ByteBuffer) throws -> Result?
+    ) rethrows -> Result? where Integer: FixedWidthInteger {
+        let prefixSize = MemoryLayout<Integer>.size
+        
+        guard let lengthPrefix = self.getInteger(at: readerIndex, endianness: endianness, as: Integer.self),
+              let messageLength = Int(exactly: lengthPrefix),
+              var messageBuffer = self.getSlice(at: readerIndex + prefixSize, length: messageLength)
+        else {
+            return nil
+        }
+        
+        guard let result = try readMessage(messageLength, &messageBuffer) else {
+            return nil
+        }
+        
+        _moveReaderIndex(forwardBy: prefixSize + messageLength)
```

Comment:
We need to the reader index forward if `readMessage` throws or returns `nil` (since we read enough bytes from the buffer).

Reply:
Done ✅